internal/engine/k8srollout/podmonitor.go

```diff
@@ -0,0 +1,160 @@
+package k8srollout
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/windmilleng/tilt/internal/k8s"
+	"github.com/windmilleng/tilt/internal/store"
+	"github.com/windmilleng/tilt/pkg/logger"
+	"github.com/windmilleng/tilt/pkg/model"
+	"github.com/windmilleng/tilt/pkg/model/logstore"
+)
+
+type PodMonitor struct {
+	pods            map[k8s.PodID]podStatus
+	trackingStarted map[k8s.PodID]bool
+}
+
+func NewPodMonitor() *PodMonitor {
+	return &PodMonitor{
+		pods:            make(map[k8s.PodID]podStatus),
+		trackingStarted: make(map[k8s.PodID]bool),
+	}
+}
+
+func (m *PodMonitor) diff(st store.RStore) []podStatus {
+	state := st.RLockState()
+	defer st.RUnlockState()
+
+	updates := make([]podStatus, 0)
+	active := make(map[k8s.PodID]bool)
+
+	for _, mt := range state.Targets() {
+		ms := mt.State
+		manifest := mt.Manifest
+		pod := ms.MostRecentPod()
+		podID := pod.PodID
+		if podID == "" {
+			continue
+		}
+
+		active[podID] = true
+
+		currentStatus := newPodStatus(pod, manifest.Name)
+		if !podStatusesEqual(currentStatus, m.pods[podID]) {
+			updates = append(updates, currentStatus)
+			m.pods[podID] = currentStatus
+		}
+	}
+
+	for key := range m.pods {
+		if !active[key] {
+			delete(m.pods, key)
+		}
+	}
+
+	return updates
+}
+
+func (m *PodMonitor) OnChange(ctx context.Context, st store.RStore) {
+	updates := m.diff(st)
+	for _, update := range updates {
+		ctx := logger.CtxWithLogHandler(ctx, podStatusWriter{
+			store:        st,
+			manifestName: update.manifestName,
+			podID:        update.podID,
+		})
+		m.print(ctx, update)
+	}
+}
+
+func (m *PodMonitor) print(ctx context.Context, update podStatus) {
+	if !m.trackingStarted[update.podID] {
+		logger.Get(ctx).Infof("Tracking new pod rollout (%s):", update.podID)
+		m.trackingStarted[update.podID] = true
+	}
+
+	m.printCondition(ctx, "Scheduled", update.scheduled, update.startTime)
+	m.printCondition(ctx, "Initialized", update.initialized, update.scheduled.LastTransitionTime.Time)
+	m.printCondition(ctx, "Ready", update.ready, update.initialized.LastTransitionTime.Time)
+}
+
+func (m *PodMonitor) printCondition(ctx context.Context, name string, cond v1.PodCondition, startTime time.Time) {
+	l := logger.Get(ctx).WithFields(logger.Fields{logger.FieldNameProgressID: name})
```

Comment:
am i right in thinking that this field updates in place?

Reply:
yep! this is the api for printing progress